internal/reminder/reminder.go

```diff
@@ -0,0 +1,197 @@
+)
+
+func init() {
+	gob.Register(map[projectProviderPair]string{})
+}
+
+// Interface is an interface over the reminder service
+type Interface interface {
+	// Start starts the reminder by sending reminders at regular intervals
+	Start(ctx context.Context) error
+
+	// Stop stops the reminder service
+	Stop()
+}
+
+// reminder sends reminders to the minder server to process entities in background.
+type reminder struct {
+	store    db.Store
+	cfg      *reminderconfig.Config
+	stop     chan struct{}
+	stopOnce sync.Once
+
+	projectListCursor string
+	repoListCursor    map[projectProviderPair]string
+
+	ticker *time.Ticker
+	logger zerolog.Logger
+
+	eventPublisher message.Publisher
+	eventDBCloser  driverCloser
+}
+
+type projectProviderPair struct {
+	// Exported for gob
+
+	ProjectId uuid.UUID
+	Provider  string
+}
+
+// NewReminder creates a new reminder instance
+func NewReminder(store db.Store, config *reminderconfig.Config) (Interface, error) {
+	level := util.ViperLogLevelToZerologLevel(config.LoggingLevel)
+	r := &reminder{
+		store:          store,
+		cfg:            config,
+		stop:           make(chan struct{}),
+		repoListCursor: make(map[projectProviderPair]string),
+		logger:         zerolog.New(os.Stdout).Level(level).With().Timestamp().Logger(),
+	}
+	err := r.restoreCursorState()
+	if err != nil {
+		// Non-fatal error, if we can't restore the cursor state, we'll start from scratch.
+		r.logger.Error().Err(err).Msg("error restoring cursor state")
+	}
+
+	pub, cl, err := r.setupSQLPublisher(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	r.eventPublisher = pub
+	r.eventDBCloser = cl
+	return r, nil
+}
+
+// Start starts the reminder by sending reminders at regular intervals
+func (r *reminder) Start(ctx context.Context) error {
+	select {
+	case <-r.stop:
+		return errors.New("reminder stopped, cannot start again")
+	default:
+	}
+
+	interval, err := time.ParseDuration(r.cfg.RecurrenceConfig.Interval)
+	if err != nil {
+		return err
+	}
+
+	if interval <= 0 {
+		return fmt.Errorf("invalid interval: %s", r.cfg.RecurrenceConfig.Interval)
+	}
+
+	r.ticker = time.NewTicker(interval)
+	defer r.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			r.logger.Info().Msg("reminder stopped")
+			return nil
+		case <-r.stop:
+			r.logger.Info().Msg("reminder stopped")
+			return nil
+		case <-r.ticker.C:
+			// In-case sending reminders i.e. iterating over entities consumes more time than the
+			// interval, the ticker will adjust the time interval or drop ticks to make up for
+			// slow receivers.
+			if errs := r.sendReminders(ctx); errs != nil {
+				for _, err := range errs {
+					r.logger.Error().Err(err).Msg("reconciliation request unsuccessful")
+				}
+			}
+		}
+	}
+}
+
+// Stop stops the reminder service
+func (r *reminder) Stop() {
+	if r.ticker != nil {
+		defer r.ticker.Stop()
+	}
+	r.stopOnce.Do(func() {
+		close(r.stop)
+		r.eventDBCloser()
+	})
+}
+
+func (r *reminder) storeCursorState() error {
```

Comment:
I don't see where this is called.  Should it be called in the loop from `Start`?

Reply:
No, this is not being used right now. This would be used by batch-building methods. `restoreCursorState()` is used and present, so it made sense to add this (for completeness and testing)